### ISSUE TITLE
HorizontalScrollSnap Mask Area doesn't work when content created dynamically

### ISSUE THREAD

**Issue by SimonDarksideJ** [Unity-UI-Extensions/com.unity.uiextensions#161]
<sup>Issue created by Андрей Коробер as [Bitbucket Issue #&#8203;125](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125) on 2017.03.11 15:28.</sup>
HorizontalScrollSnap.AddChild method doesn't add item to ScrollSnapBase.ChildObjects, and ScrollSnapBase.UpdateVisible() doesn't update item's visibility, because it has check:

```C#
   internal void UpdateVisible()
        {
            //If there are no objects in the scene or a mask, exit
            if (!MaskArea || ChildObjects == null || ChildObjects.Length < 1 || _screensContainer.childCount < 1)
            {
                return;
            }
```

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.03.12 10:47](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-35027099), @SimonDarksideJ commented:</sup>
Can you give me a working sample / use case where this isn't working/ To ensure the fix meets your need?

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.04.13 13:16](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-35948222), @SimonDarksideJ commented:</sup>
Should be fixed in source, can you check @SORM ?

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.04.19 14:59](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-36070464), frima_kb commented:</sup>
I had to use AddChild and it seems to be working as expected.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.06.05 17:23](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-37382103), @SimonDarksideJ commented:</sup>
Can this issue be closed then @frima_kb ?

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.06.05 18:04](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-37382795), frima_kb commented:</sup>
I think so :D

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.06.05 18:53](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-37383588), @SimonDarksideJ commented:</sup>
Closed on user request

**Comment by SimonDarksideJ** [comment]
<sup>On 2017.06.05 18:53 @SimonDarksideJ modified issue:</sup>
**status** changed `new` → `resolved`

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.07.31 02:51](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-38735480), Kurt Jing commented:</sup>
@frima_kb I still have this bug, and can not see in the source where the fix is. HorizontalScrollSnap.AddChild does not update childobjects, and childobjects.length will less 1

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.07.31 08:35](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-38739396), @SimonDarksideJ commented:</sup>
Do you have a sample to demonstrate your issue @kurtsea129 ? Or a set of replication steps to set it up so I can test.

The AddChild methods start at line 97 in the HSS script
https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/src/974f906fb0b3e401bf2256e73f1e564fbe6374be/Scripts/Layout/HorizontalScrollSnap.cs?at=master&fileviewer=file-view-default#HorizontalScrollSnap.cs-99

You need to add children programmatically using the AddChild methods, not by adding children manually through the hierarchy as that is only done on start.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.08.01 01:16](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-38802209), Kurt Jing commented:</sup>
Hi @ddreaper , to dynamically add items like below


```c#

for (int i = 0; i < 10; i++)
 {
            GameObject item= GameObject.Instantiate(itemPrefab);
            scrolSnap.AddChild(item);
  }
```

If you look through the code of AddChild, it does not update ScrollSnapBase.ChildObjects and call UpdateVisible, but as Andrey mentioned in UpdateVisible function it checks length of ChildObjects which will always be 0 since children added dynamically by AddChild. So UpdateVisible will just return and does nothing, that is why Mask won't work.

To fixed that, currently I have to do something like


```C#

for (int i = 0; i < 10; i++)
 {
            GameObject item= GameObject.Instantiate(itemPrefab);
            scrolSnap.ChildObjects[i] = item;
            scrolSnap.AddChild(item);
  }
```

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.08.07 09:53](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-38931239), @SimonDarksideJ commented:</sup>
Related issue and possible fix to investigate

**Comment by SimonDarksideJ** [comment]
<sup>On 2017.08.07 09:53 @SimonDarksideJ modified issue:</sup>
**status** changed `resolved` → `open`

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.08.07 11:48](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-38934002), @SimonDarksideJ commented:</sup>
@kurtsea129 The AddChild methods for both Vertical and Horizontal ScrollSnap both have a call to "InitialiseChildObjectsFromScene()"
This method completely regenerates the ChildObjects array when a new child is added.

I'll do some more tests, but the test scene does have this example as part of it's base tests.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.08.07 11:52](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-38934070), @SimonDarksideJ commented:</sup>
*Edit, scratch that.  Just noticed the Horizontal SS AddPage method is NOT calling InitialiseChildObjectsFromScene. GAH.
Fixing that now :D

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.08.07 11:54](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/125#comment-38934099), @SimonDarksideJ commented:</sup>
Horizontal Scroll Snap "AddChild" method was NOT calling "InitialiseChildObjectsFromScene()", added

Resolves #125


Referenced by cdef4bdfefb4

**Comment by SimonDarksideJ** [comment]
<sup>On 2017.08.07 11:54 @SimonDarksideJ modified issue:</sup>
**status** changed `open` → `resolved`
